src/hooks/useNotificationModal.ts

```diff
@@ -0,0 +1,204 @@
+import {
+  SaveSubscriptionMutation,
+  SaveSubscriptionMutationVariables,
+} from "gql/generated/types";
+import { useBannerDispatchContext } from "context/banners";
+import { useState, useEffect } from "react";
+import { useMutation } from "@apollo/react-hooks";
+import { SAVE_SUBSCRIPTION } from "gql/mutations/save-subscription";
+import get from "lodash.get";
+
+export interface UseNotificationModalProps {
+  subscriptionMethodControls: SubscriptionMethods;
+  triggers: Trigger[];
+  resourceType: ResourceType;
+  resourceId: string;
+  onCancel: () => void;
+}
+export const useNotificationModal = ({
+  triggers,
+  subscriptionMethodControls,
+  resourceType,
+  resourceId,
+  onCancel,
+}: UseNotificationModalProps) => {
+  const dispatchBanner = useBannerDispatchContext();
+  const [selectedSubscriptionMethod, setSelectedSubscriptionMethod] = useState(
+    ""
+  );
+  // target represents the input value for a subscription method
+  const [target, setTarget] = useState<Target>({});
+  const [isFormValid, setIsFormValid] = useState(false);
+  const [extraFieldErrorMessages, setExtraFieldErrorMessages] = useState<
+    string[]
+  >([]);
+  const [saveSubscription, { loading: mutationLoading }] = useMutation<
+    SaveSubscriptionMutation,
+    SaveSubscriptionMutationVariables
+  >(SAVE_SUBSCRIPTION, {
+    onCompleted: () => {
+      dispatchBanner.successBanner("Your subscription has been added");
+    },
+    onError: (err) => {
+      dispatchBanner.errorBanner(
+        `Error adding your subscription: '${err.message}'`
+      );
+      // TODO: save error
+    },
+  });
+  const [selectedTriggerId, setSelectedTriggerId] = useState<string>("");
+  const [extraFieldInputVals, setExtraFieldInputVals] = useState<
+    ExtraFieldInputVals
+  >({});
+
+  // extraFields represents schema additional inputs required for the selected trigger
+  const { extraFields } =
+    triggers.find(({ trigger }) => trigger === selectedTriggerId) ?? {};
+
+  // clear the input vals for the extraFields when the extraFields change
+  useEffect(() => {
+    setExtraFieldInputVals(
+      (extraFields ?? []).reduce(clearExtraFieldsInputCb, {})
```

Comment:
not necessary to reduce on an empty array.

Reply:
if extraFields evaluates to false I would like to setExtraInputVals to an empty object so I think this is fine here